noxfile.py

```diff
@@ -0,0 +1,166 @@
+    """Run mypy on cachi2 and tests directories and noxfile.py."""
+    install_requirements(session)
+    cmd = "mypy --install-types --non-interactive cachi2 tests noxfile.py"
+    session.run(*cmd.split(), *session.posargs, silent=True)
+
+
+@nox.session(name="unit-tests-python", python=parse_supported_python_versions())
+def unit_tests(session: Session) -> None:
+    """Run unit tests and generate coverage report."""
+    install_requirements(session)
+    cmd = "pytest --log-level=DEBUG --cov=cachi2 --cov-config=pyproject.toml --cov-report=term --cov-report=html --no-cov-on-fail tests/unit"
+    session.run(*cmd.split(), *session.posargs)
+
+
+@nox.session(name="coverage-html")
+def coverage_html(session: Session) -> None:
+    """Open the coverage report in the browser."""
+    session.install("coverage")
+    session.run("coverage", "html")
+    session.run("xdg-open", "htmlcov/index.html", external=True)
+
+
+def _run_integration_tests(session: Session, env: dict[str, str]) -> None:
+    install_requirements(session)
+    netrc = "machine 127.0.0.1 login cachi2-user password cachi2-pass"
+    default_env = {"CACHI2_TEST_NETRC_CONTENT": os.getenv("CACHI2_TEST_NETRC_CONTENT", netrc)}
+    default_env.update(env)
+    cmd = "pytest --log-cli-level=WARNING tests/integration"
+    session.run(*cmd.split(), *session.posargs, env=default_env)
+
+
+@nox.session(name="integration-tests")
+def integration_tests(session: Session) -> None:
+    """Run integration tests only for the affected code base in the current branch."""
+    _run_integration_tests(session, {})
+
+
+@nox.session(name="all-integration-tests")
+def all_integration_tests(session: Session) -> None:
+    """Run all integration tests that are available."""
+    _run_integration_tests(
+        session,
+        {
+            "CACHI2_RUN_ALL_INTEGRATION_TESTS": "true",
+            "CACHI2_TEST_LOCAL_PYPISERVER": "true",
+            "CACHI2_TEST_LOCAL_DNF_SERVER": "true",
+        },
+    )
+
+
+@nox.session(name="generate-test-data")
+def generate_test_data(session: Session) -> None:
+    """Run all integration tests that are available and update SBOMs."""
+    _run_integration_tests(
+        session,
+        {
+            "CACHI2_GENERATE_TEST_DATA": "true",
+            "PYTEST_ADDOPTS": "-k test_e2e",
+        },
+    )
+
+
+@nox.session(name="pip-compile")
+def pip_compile(session: Session) -> None:
+    """Update requirements.txt and requirements-extras.txt files."""
+    PWD = session.env["PWD"]
+    # git must be installed in the image due to setuptools-scm that has it as a direct dependency
+    pip_compile_cmd = """
+    apk add git && \
+    pip3 install pip-tools && \
+    pip-compile \
+        --generate-hashes \
+        --output-file=requirements.txt \
+        --rebuild \
+        pyproject.toml && \
+    pip-compile \
+        --all-extras \
+        --generate-hashes \
+        --output-file=requirements-extras.txt \
+        --rebuild \
+        pyproject.toml
```

Comment:
Has this actually worked for you? It didn't when I executed it locally, some '&&' stringifying happening that prevented the shell inside the container to actually run the commands properly.

What eventually worked for me though was dumping the commands to a temporary script and mounting that to the container:
```    tempfile = None
    with NamedTemporaryFile(mode="w", prefix="cachi2_pipcompile", suffix=".sh", delete=False) as tf:
        tempfile = tf.name
        deps_cmds = """
        #!/bin/sh
        # git must be installed in the image due to setuptools-scm that has it as a direct dependency
        apk add --no-cache git
        pip3 install pip-tools
        """

        pip_compile_cmds = """
        pip-compile \
            --generate-hashes \
            --output-file=requirements.txt \
            --rebuild \
            pyproject.toml
        pip-compile \
            --all-extras \
            --generate-hashes \
            --output-file=requirements-extras.txt \
            --rebuild \
            pyproject.toml
        """
        tf.write(deps_cmds + pip_compile_cmds)

    target = Path(tempfile).name
    cmd = f"podman run --rm --volume {PWD}:/cachi2:rw,Z --volume {tf.name}:/{target}:ro,Z --workdir /cachi2 docker.io/library/python:3.9-alpine"
    session.run(*cmd.split(), *["sh", f"/{target}"], external=True)
    os.unlink(tempfile)
```

I didn't bother with `dedent`, but it could be incorporated to make the temp script nicer for debugging purposes. The point being, something along ^those lines should be done instead.